File: day05/naughty.c
#include <stdio.h>
#include <string.h>

#define MAX_LEN 100

int main(int argc, char *argv[]) {
        FILE *fp;
        char line[MAX_LEN];
        int num_nice = 0;

        if (argc != 2) {
                fprintf(stderr, "Usage: naughty <file>\n");
                return 1;
        }

        if ((fp = fopen(argv[1], "r")) == NULL) {
                perror(argv[1]);
                return 1;
        }

        while (fgets(line, MAX_LEN, fp) != NULL) {
                int vowels;
                int doub;
                int special;
                int i, j, k;
                size_t len = strlen(line);

                /* look for 3 vowels */
                int num_vowels = 0;
                for (i = 0; i < len; i++) {
                        char c = line[i];
                        if (c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u') {
                                num_vowels++;
                                if (num_vowels >= 3)
                                        break;
                        }
                }
                vowels = (num_vowels >= 3) ? 1 : 0;

                /* look for double letters */
                doub = 0;
                for (i = 0; !doub && i < len - 1; i++) {
                        if (line[i] == line[i+1])
                                doub = 1;
                }

                /* look for special strings */
                special = 0;
                for (i = 0; !special && i < len - 1; i++) {
                        if (strncmp(line+i, "ab", 2) == 0 ||
                            strncmp(line+i, "cd", 2) == 0 ||
                            strncmp(line+i, "pq", 2) == 0 ||
                            strncmp(line+i, "xy", 2) == 0) 
                                special = 1;
                }

                if (vowels && doub && !special) {
                        printf("nice    %s", line);
                        num_nice++;
                } else {
                        printf("naughty %s", line);
                }
        }

        printf("%d were nice\n", num_nice);

        return 0;
}

File: day05/naughty.pl
#!/usr/bin/env perl
use strict;
use warnings;
use 5.10.0;

my $num_nice = 0;
while (<>) {
    chomp;
    my $vowels;
    my $double;
    my $special;

    if (/[aeiou].*[aeiou].*[aeiou]/) {
        $vowels = 1;
    } else {
        $vowels = 0;
    }

    if (/(.)\1/) {
        $double = 1;
    } else {
        $double = 0;
    }

    if (/ab|cd|pq|xy/) {
        $special = 1;
    } else {
        $special = 0;
    }

    if ($vowels && $double && !$special) {
        say "nice    $_";
        $num_nice++;
    } else {
        say "naughty $_";
    }
}

say "$num_nice were nice";
